### ISSUE TITLE
Not finding global and data frame variables in functions

### ISSUE THREAD

**Issue by fredcallaway** [r-lib/lintr#2691]
Apologies if this is documented behavior but I couldn't find it (and it does seem incorrect to me).

```r
library(tidyverse)
mutate(cars, speed = speed * 10)

foo <- function(x) {
    mutate(x, speed = speed * 10)
}
```

This produces two warnings:
```
...tmp.r:5:5: warning: [object_usage_linter] no visible global function definition for 'mutate'
    mutate(x, speed = speed * 10)
    ^~~~~~
...tmp.r:5:23: warning: [object_usage_linter] no visible binding for global variable 'speed'
    mutate(x, speed = speed * 10)
                      ^~~~~
```

**Comment by MichaelChirico** [comment]
Could you give the exact code on how you are running `lint()`? I'd expect:

 1. `mutate()` would be found
 2. `speed` would not

The latter is extremely difficult to get right statically; building logic for that is out of scope. Eventually, declarations will be needed.

Can you also clarify if this is R script code, or package code?

**Comment by MichaelChirico** [comment]
Here's two ways to make the `mutate()` lint go away:

 1. Running `library(tidyverse)` in your session before `lint()`
 2. Changing `library(tidyverse)` to `library(dplyr)`

If you're wondering why the first `mutate()` (under `library()`) does not generate a lint while the second does, it's because `object_usage_linter()` only looks inside functions.

> As for the speed case, I think a reasonable default would be to disable missing variable warnings inside dplyr calls.

Sorry, but this will generate tons of false positives as well. The problem here is a lot messier & more subtle than you might realize. I have extensive experience with this, and there is a way to make most of the false positives go away, but it is pretty hacky, I would be hesitant to ship it with {lintr}.

> is there a way I can globally disable the missing function/variable warnings?

My advice is to just disable `object_usage_linter()` for scripts: `linters_with_tags("correctness", object_usage_linter=NULL)`.

I filed #2696 which would partially ameliorate your issue; this has come up before as well: #1127.
